packages/descendants/src/Highlight/README.md

```diff
@@ -0,0 +1,115 @@
+# Highlight Management
+
+A suite of utilities for managing the highlight state of Descendants.
+
+# Usage
+
+```tsx
+/** HighlightDemo.tsx */
+
+const HighlightDemo = () => {
+  // Initialize a new Descendants object
+  const { getDescendants, dispatch } = useInitDescendants<HTMLDivElement>();
+  // Initialize a new highlight state
+  const { highlight, moveHighlight, setHighlight } =
+    useHighlight(getDescendants);
+
+  return (
+    // Init the Descendants context provider
+    <DescendantsProvider
+      context={TestDescendantContext}
+      descendants={getDescendants()}
+      dispatch={dispatch}
+    >
+      {/* Init the Highlight context provider */}
+      <HighlightProvider
+        context={TestHighlightContext}
+        highlight={highlight}
+        setHighlight={setHighlight}
+      >
+        {items.map(item => (
+          <HighlightItem key={item}>{item}</HighlightItem>
+        ))}
+      </HighlightProvider>
+    </DescendantsProvider>
+  );
+};
+```
+
+```tsx
+/** HighlightItem.tsx */
+
+const HighlightItem = ({ children }: PropsWithChildren<{}>) => {
+  const { ref, id } = useDescendant(TestDescendantContext);
+  const { highlight } = useHighlightContext(TestHighlightContext);
+
+  return (
+    <div
+      ref={ref}
+      style={{
+        color: isHighlighted ? 'red' : 'black',
+      }}
+    >
+      {children}
+    </div>
+  );
+};
+```
+
+## `options.onInit`: Setting the initial highlight
+
+To set the initial highlight (or run side-effects once on init), provide an `onInit` callback to the `options` argument of `useHighlight`.
+
+```ts
+const { getDescendants, dispatch } = useInitDescendants<HTMLDivElement>();
+const { highlight, setHighlight } = useHighlight(getDescendants, {
+  // Fired once when the descendants are initialized
+  onInit: () => {
+    setHighlight(0);
+  },
+});
+```
+
+## `options.onChange`: Responding to Highlight changes
+
+To run side-effects after the highlight has been updated, provide an `onChange` callback to the `options` argument of `useHighlight`.
+
+```ts
+const { highlight, setHighlight } = useHighlight(getDescendants, {
+  // Fired when the highlight state changes
+  onChange: (next: Descendant | undefined) => {
+    if (next) {
+      // Focus the element when highlight is updated
+      next.ref.current.focus();
+    }
+  },
+});
+```
+
+## `options.filter`: Filtering descendants
+
+Occasionally some descendants should not be able to be highlighted. In these cases, provide a `filter` function to the `options` argument of `useHighlight`.
+
+```ts
+const { highlight, setHighlight } = useHighlight(getDescendants, {
+  // Exclude descendants with the `disabled` prop from receiving highlight
+  filter: (descendant: Descendant) => !descendant.props.disabled,
```

Comment:
out of curiosity... why can't we do 
```
useHighlight(getDescendants.filter(d => !d.disabled)
```

Reply:
Because then we don't have the correct total number of descendants to compute when to overflow

Reply:
wait why not?

Reply:
it's relative to the number of enabled indexes not to the number of total elements
you just `% enabledIndexes`

Reply:
the overflow doesn't care about the total it's only concerned with the total number that can receive a highlight

Reply:
<img width="254" alt="Screenshot 2024-07-10 at 6 19 08 PM" src="https://github.com/mongodb/leafygreen-ui/assets/2414030/db1f7c95-ca54-48f7-a466-6ad517edb605">

Short answer: because the `index` is an attribute on a `Descendant` object, and  if we filter these, the indexes won't be aligned.

If I call `moveHighlight(1)` in the case above, we want it to skip from `Eliane` to `George`. Currently we do this with the following logic:
- The current `highlight` object has `highlight.index == 4` 
- Increment this index by `1` (`i = 5`)
- Check if the descendant with the incremented index is valid (index 5 "Fred" is disabled)
- Keep incrementing by 1 until we find a non-disabled one (`i = 6`)
- Set the highlight to the element with index `6` ("George")

Reply:
If we filter the array, we can get into a case where the `index` property on the `highlight` object doesn't match the index in the array. 
Consider the inverse case:
- "George" is selected (`index === 6`)
- Call `moveHighlight(-1)`
- Check if `descendants[5]` is valid
- If not, check `descendants[4]` (etc.)
- Set the highlight to `descendants[4]` (`"Eliane"`)

If we were to pre-filter the descendants list, then we would decrement the index of "George" (`6`) down to `5`, and then check `filteredDescendants[5]`, which would still be "George"

Reply:
A potential alternative if we really don't want this `options.filter`:

We expect a pre-filtered list of descendants. 
To move the highlight relative to the current one we 
- Loop through the `filteredDescendants` array to find the new index of the current element
- update this index
- return the `filteredDescendant[x]`

Not a lot of effort, but it includes a redundant search through an array, as well as a `filter` loop every render

Reply:
I keep typing and resetting - I think what you have is OK to move forward with, we can get into the nuances offline.

I think there's a lot of assumptions we're baking in by colocating these two contexts, which are discrete and completely independent but become mentally associated by this colocation. 

For example: the moveHighlight function gets complicated if you're being asked to keep track of two different index values in a given component (one from descendants, one from highlihgt) but theoretically if these weren't packaged together you could easily return an index from this context.

All this to say - comfortable going forward with as-is, but an interesting discussion to reason about in my head :)

Reply:
Let's chat tomorrow. 

Note, the arg passed into `moveHighlight` is _relative to the current_ highlight, not an absolute index. 

To set an absolute index, you'd use `setHighlight`. 

I decided that integrating the relative updating into the utility was better than having the consumer compute that externally